mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java

```diff
@@ -206,27 +209,26 @@ public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) throws MailboxExcep
             .orElse(null);
     }
 
-    @Override
     public Map<MessageUid, MessageMetaData> expungeMarkedForDeletionInMailbox(Mailbox mailbox, MessageRange messageRange) throws MailboxException {
         CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
 
-        CompletableFuture<Stream<SimpleMailboxMessage>> messagesToDelete =
-            CompletableFutureUtil.flatMapOnAll(
-                deletedMessageDAO.retrieveDeletedMessage(mailboxId, messageRange),
+        return CompletableFutureUtil.chainOnStream(deletedMessageDAO.retrieveDeletedMessage(mailboxId, messageRange))
+            .thenComposeOnAll(
                 messageId ->
                     messageIdDAO.retrieve(mailboxId, messageId))
-
-            .thenApply(ids -> ids
-                .flatMap(OptionalConverter::toStream)
-                .collect(Guavate.toImmutableList()))
-
-            .thenCompose(ids -> retrieveMessages(ids, FetchType.Metadata, Optional.empty()));
-
-        return CompletableFutureUtil.performOnAll(messagesToDelete, message ->
+            .flatMap(OptionalConverter::toStream)
+            .thenCompose(ids ->
+                retrieveMessages(
+                    ids.collect(Guavate.toImmutableList()),
+                    FetchType.Metadata,
+                    Optional.empty())
+            )
+            .performOnAll(message ->
                 deleteAsFuture(message, mailboxId))
```

Comment:
Now it' short enought to directly put deleteAsFuture after -> and still get nice display in Github